cnxdb/migrations/20200506144128_update_trees_slug_values.py

```diff
@@ -0,0 +1,196 @@
+# -*- coding: utf-8 -*-
+
+import os
+import time
+from datetime import timedelta
+from functools import wraps
+
+import requests
+from dbmigrator import deferred
+from dbmigrator import logger
+from cnxcommon.urlslug import generate_slug
+
+
+ARCHIVE_DOMAIN = os.getenv('ARCHIVE_DOMAIN')
+VARNISH_HOSTS = os.getenv('VARNISH_HOSTS', '').split(',')
+VARNISH_PORT = int(os.getenv('VARNISH_PORT', 80))
+
+
+BATCH_SIZE = 1000
+TREE_QUERY = """\
+WITH RECURSIVE t(node, title, path, value, depth, corder, is_collated) AS (
+    SELECT nodeid, ARRAY[title], ARRAY[nodeid], documentid, 1, ARRAY[childorder],
+           is_collated
+    FROM trees tr, modules m
+    WHERE tr.documentid = m.module_ident
+          AND
+          tr.parent_id IS NULL
+          AND
+          tr.is_collated = TRUE
+UNION ALL
+    /* Recursion */
+    SELECT c1.nodeid,
+           /* concat the new record to the hierarchy of titles */
+           t.title || ARRAY[c1.title],
+           /* concat the new record to the path */
+           t.path || ARRAY[c1.nodeid],
+           c1.documentid,
+           t.depth+1,
+           t.corder || ARRAY[c1.childorder],
+           c1.is_collated
+    FROM trees c1 JOIN t ON (c1.parent_id = t.node)
+    WHERE NOT nodeid = ANY (t.path)
+          AND
+          t.is_collated = c1.is_collated
+)
+SELECT node, title
+FROM t LEFT JOIN modules m ON t.value = m.module_ident
+WINDOW w AS (ORDER BY corder)
+ORDER BY corder
+;
+"""
+
+# https://wiki.postgresql.org/wiki/Unnest_multidimensional_array
+CREATE_REDUCE_DIM = """\
+CREATE OR REPLACE FUNCTION public.reduce_dim(anyarray)
+RETURNS SETOF anyarray AS
+$function$
+DECLARE
+    s $1%TYPE;
+BEGIN
+    FOREACH s SLICE 1  IN ARRAY $1 LOOP
+        RETURN NEXT s;
+    END LOOP;
+    RETURN;
+END;
+$function$
+LANGUAGE plpgsql IMMUTABLE;
+"""
+
+DROP_REDUCE_DIM = """\
+DROP FUNCTION public.reduce_dim(anyarray);
+"""
+
+UPDATE_STMT = """\
+UPDATE trees SET slug = q.slug
+from (
+  SELECT yyy[1] AS id, yyy[2] AS slug FROM reduce_dim(%s) AS yyy
+) AS q
+WHERE nodeid = q.id::int
+;
+"""
+
+
+def batcher(seq, size):
+    for pos in range(0, len(seq), size):
+        yield seq[pos:pos + size]
+
+
+def generate_update_values(nodeid, title):
+    """Returns a sequence of trees.nodeid and trees.slug
+    to be used to update the trees slug table value.
+
+    """
+    logger.info("processing... {} - {}".format(nodeid, title))
+    try:
+        slug = generate_slug(*title)
```

Comment:
Since my current content doesn't have `os-part-text`, I modified this when testing locally to be: `generate_slug(*title) + "test"` and confirmed:
1. Slugs were updated accordingly (e.g. what they were + the suffix)
2. The migration didn't run twice